codegenerator/model/api.go

```diff
@@ -132,67 +135,75 @@ func (auth *Auth) apiCall(payload interface{}, method, route string, result inte
-		// not sure if we need to regenerate this with each call, will leave in here for now...
-		credentials := &hawk.Credentials{
-			ID:   auth.ClientId,
-			Key:  auth.AccessToken,
-			Hash: sha256.New,
-		}
-		reqAuth := hawk.NewRequestAuth(httpRequest, credentials, 0).RequestHeader()
-		httpRequest.Header.Set("Authorization", reqAuth)
-	}
 	httpRequest.Header.Set("Content-Type", "application/json")
 	callSummary.HttpRequest = httpRequest
 	httpClient := &http.Client{}
-	// fmt.Println("Request\n=======")
-	// fullRequest, err := httputil.DumpRequestOut(httpRequest, true)
-	// fmt.Println(string(fullRequest))
-	response, err := httpClient.Do(httpRequest)
-	// fmt.Println("Response\n========")
-	// fullResponse, err := httputil.DumpResponse(response, true)
-	// fmt.Println(string(fullResponse))
-	if err != nil {
-		callSummary.Error = err
-		return result, callSummary
-	}
-	defer response.Body.Close()
-	callSummary.HttpResponse = response
-	// now read response into memory, so that we can return the body
-	body, err := ioutil.ReadAll(response.Body)
-	if err != nil {
-		callSummary.Error = err
-		return result, callSummary
-	}
-	// response.Body = ioutil.NopCloser(strings.NewReader(string(body)))
-	callSummary.HttpResponseBody = string(body)
 
-	// now check if http response code isn't in good range [200,300)...
-	if respCode := response.StatusCode; respCode/100 != 2 {
-		callSummary.Error = BadHttpResponseCode{
-			HttpResponseCode: respCode,
-			Message: "HTTP response code " + strconv.Itoa(respCode) + " to http " +
-				method + " request to " + auth.BaseURL + route + ":\n" +
-				string(body) + "\n",
+	// function to perform http request - we call this using backoff library to
+	// have exponential backoff in case of intermittent failures (e.g. network
+	// blips)
+	httpCall := func() error {
+		log.Printf("Making http reqest: %s\n", httpRequest)
+		// Refresh Authorization header with each call...
+		// Only authenticate if client library user wishes to.
+		if auth.Authenticate {
+			credentials := &hawk.Credentials{
+				ID:   auth.ClientId,
+				Key:  auth.AccessToken,
+				Hash: sha256.New,
+			}
+			reqAuth := hawk.NewRequestAuth(httpRequest, credentials, 0).RequestHeader()
+			httpRequest.Header.Set("Authorization", reqAuth)
 		}
-		return result, callSummary
-	}
-	// if result is nil, it means there is no response body json
-	if reflect.ValueOf(result).IsValid() && !reflect.ValueOf(result).IsNil() {
-		err := json.Unmarshal(body, &result)
+		response, err := httpClient.Do(httpRequest)
 		if err != nil {
-			callSummary.Error = err
-			return result, callSummary
+			return err
 		}
+		callSummary.HttpResponse = response
+		defer response.Body.Close()
+		// now read response into memory, so that we can return the body
+		body, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return err
+		}
+		callSummary.HttpResponseBody = string(body)
+
+		// now check if http response code isn't in good range [200,300)...
```

Comment:
Do all of this outside the retry loop, e.g. create `BadHttpResponseCode` and parsing JSON... The only thing want to check here is the response code, if it's 5xx then return an error... Otherwise return nothing and parse the response code + parse JSON after the retry loop.